src/raftstore/store/worker/region.rs

```diff
@@ -631,4 +802,149 @@ mod test {
+            // construct snapshot
+            let (tx, rx) = mpsc::sync_channel(1);
+            sched
+                .schedule(Task::Gen {
+                    region_id: id,
+                    notifier: tx,
+                })
+                .unwrap();
+            let s1 = rx.recv().unwrap();
+            let data = s1.get_data();
+            let key = SnapKey::from_snap(&s1).unwrap();
+            let mgr = SnapManager::new(snap_dir.path().to_str().unwrap(), None);
+            let mut s2 = mgr.get_snapshot_for_sending(&key).unwrap();
+            let mut s3 = mgr.get_snapshot_for_receiving(&key, &data[..]).unwrap();
+            io::copy(&mut s2, &mut s3).unwrap();
+            s3.save().unwrap();
+
+            // set applying state
+            let wb = WriteBatch::new();
+            let handle = db.cf_handle(CF_RAFT).unwrap();
+            let region_key = keys::region_state_key(id);
+            let mut region_state = db
+                .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_key)
+                .unwrap()
+                .unwrap();
+            region_state.set_state(PeerState::Applying);
+            wb.put_msg_cf(handle, &region_key, &region_state).unwrap();
+            db.write(wb).unwrap();
+
+            // apply snapshot
+            let status = Arc::new(AtomicUsize::new(JOB_STATUS_PENDING));
+            sched
+                .schedule(Task::Apply {
+                    region_id: id,
+                    status,
+                })
+                .unwrap();
+        };
+        let wait_apply_finish = |id: u64| {
+            let region_key = keys::region_state_key(id);
+            loop {
+                thread::sleep(Duration::from_millis(100));
+                if db
+                    .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_key)
+                    .unwrap()
+                    .unwrap()
+                    .get_state() == PeerState::Normal
+                {
+                    break;
+                }
+            }
+        };
+        let cf = db.cf_handle(CF_DEFAULT).unwrap();
+
+        // snapshot will not ingest cause already write stall
+        gen_and_apply_snap(1);
+        assert_eq!(rocksdb::get_cf_num_files_at_level(&db, cf, 0).unwrap(), 6);
+
+        // compact all files to the bottomest level
+        rocksdb::compact_files_in_range(&db, None, None, None).unwrap();
+        assert_eq!(rocksdb::get_cf_num_files_at_level(&db, cf, 0).unwrap(), 0);
+
+        wait_apply_finish(1);
+
+        // the pending apply task should be finished and snapshots are ingested.
+        // note that when ingest sst, it may flush memtable if overlap,
+        // so here will two level 0 files.
+        assert_eq!(rocksdb::get_cf_num_files_at_level(&db, cf, 0).unwrap(), 2);
+
+        // no write stall, ingest without delay
+        gen_and_apply_snap(2);
+        wait_apply_finish(2);
+        assert_eq!(rocksdb::get_cf_num_files_at_level(&db, cf, 0).unwrap(), 4);
```

Comment:
I think it is better to use pending_applies.len() for verification here.

Reply:
I has attempted to use that, it is hard to get pending_applies.len() outerly